taqueria-vscode-extension-web-ui/src/data-editors/MichelineEditor.tsx

```diff
@@ -0,0 +1,236 @@
+
+export type OriginalFormat = 'micheline' | 'json';
+
+export type MichelineValue = {
+	originalFormat: 'micheline';
+	value: string;
+} | {
+	originalFormat: 'json';
+	value: object;
+};
+
+const json2Micheline = (v: object) => {
+	let michelson: string | undefined = undefined;
+	try {
+		const expression = parser.parseJSON(v);
+		michelson = emitMicheline(expression);
+		return michelson;
+	} catch (e: any) {
+		return `${e}`;
+	}
+};
+
+const micheline2Json = (v: string) => {
+	let json: object | undefined | null = undefined;
+	try {
+		json = parser.parseMichelineExpression(v);
+		return json;
+	} catch (e: any) {
+		return undefined;
+	}
+};
+
+export const getMicheline = (value: MichelineValue | undefined) => {
+	if (!value) {
+		return undefined;
+	}
+	if (value.originalFormat === 'micheline') {
+		return value.value;
+	}
+	return json2Micheline(value.value);
+};
+
+export const getJson = (value: MichelineValue | undefined) => {
+	if (!value) {
+		return undefined;
+	}
+	if (value.originalFormat === 'json') {
+		return value.value;
+	}
+	return micheline2Json(value.value);
+};
+
+export const getFriendlyDataType = (dataType: any): string => {
+	if (!dataType || typeof dataType != 'object' || !dataType.prim) {
+		return `Error: ${JSON.stringify(dataType)}`;
+	}
+	let friendlyDataType: string;
+	switch (dataType.prim) {
+		case 'list':
+		case 'set':
+		case 'option':
+			friendlyDataType = `${dataType.prim} of ${getFriendlyDataType(dataType.args[0])}`;
+			break;
+		case 'unit':
+			return 'unit';
+		case 'string':
+		case 'nat':
+		case 'int':
+		case 'bytes':
+		case 'timestamp':
+		case 'mutez':
+		case 'address':
+		case 'key':
+		case 'key_hash':
+		case 'signature':
+		case 'chain_id':
+		case 'bls12_381_g1':
+		case 'bls12_381_g2':
+		case 'bls12_381_fr':
+		case 'sapling_transaction ms':
+		case 'sapling_state ms':
+		case 'ticket':
+		case 'chest':
+		case 'chest_key':
+		case 'tx_rollup_l2_address':
+		case 'bool':
+			friendlyDataType = dataType.prim;
+			break;
+		case 'pair':
+			friendlyDataType = `${dataType.prim} of ${(dataType.args as any[]).map(x => getFriendlyDataType(x)).join(', ')}`;
+			break;
+		case 'map':
+		case 'big_map':
+			friendlyDataType = `${dataType.prim} from ${getFriendlyDataType(dataType.args[0])} to ${
+				getFriendlyDataType(dataType.args[1])
+			}`;
+			break;
+		default:
+			return `Error: ${JSON.stringify(dataType)}`;
+	}
+	if (dataType.annots && dataType.annots.length) {
+		return `${(dataType.annots as string[]).map(x => x?.substring(1)).join(' ')} (${friendlyDataType})`;
+	}
+	return friendlyDataType;
+};
+
+declare const vscodeApi: WebviewApi<any>;
+
+export const MichelineEditor = (
+	{ input: { dataType, value, actionTitle, showDiagnostics }, onMessage }: {
+		input: { dataType: any; value?: MichelineValue; actionTitle?: string; showDiagnostics?: boolean };
+		onMessage: MichelineEditorMessageHandler;
+	},
+) => {
+	const previousState = vscodeApi.getState() as MichelineValue | undefined;
```

Comment:
Our code styling isn't consistent with the approach here.

We should decide whether to use:
```
const previousState: MichelineValue | undefined = vscodeApi.getState()
```
or

```
const previousState = vscodeApi.getState() as MichelineValue | undefined;
```

I think the consistency will allow code reviewers to work faster.

Reply:
I personally just prefer: `const previousState = vscodeApi.getState()` and let the compiler infer as much as possible to avoid having to update/maintain type annotations.

Reply:
@mweichert `.getState()` returns `any`, that's why there's type casting.